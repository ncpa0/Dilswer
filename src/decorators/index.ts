export { ClassWithValidation, validate } from "./validation-decorator";
