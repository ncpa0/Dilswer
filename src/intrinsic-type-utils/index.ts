export * from "@Intrinsic/and";
export * from "@Intrinsic/exclude";
export * from "@Intrinsic/omit";
export * from "@Intrinsic/partial";
export * from "@Intrinsic/pick";
export * from "@Intrinsic/required";
