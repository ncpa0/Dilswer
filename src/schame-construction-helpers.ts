import { AllDataTypes } from ".";
import type {
  ArrayOf,
  Enum,
  EnumMember,
  Literal,
  OneOf,
  RecordOf,
  SetOf,
  TypeSchema,
} from "./types";

export const BasicDataTypes = {
  Unknown: "unknown",
  String: "string",
  Number: "number",
  Boolean: "boolean",
  Symbol: "symbol",
  Function: "function",
  Null: "null",
  Undefined: "undefined",
} as const;

export const DataType = {
  ...BasicDataTypes,
  RecordOf<TS extends TypeSchema>(args: TS): RecordOf<TS> {
    return { recordOf: args };
  },
  ArrayOf<DT extends AllDataTypes[]>(...args: DT): ArrayOf<DT> {
    return { arrayOf: args };
  },
  SetOf<DT extends AllDataTypes[]>(...args: DT): SetOf<DT> {
    return { setOf: args };
  },
  OneOf<DT extends AllDataTypes[]>(...args: DT): OneOf<DT> {
    return { oneOf: args };
  },
  Literal<V extends string | number | boolean>(value: V): Literal<V> {
    return { literal: value };
  },
  EnumMember<M extends number | string>(enumMember: M): EnumMember<M> {
    return { enumMember };
  },
  Enum<T extends string, TEnumValue extends string | number>(enumInstance: {
    [key in T]: TEnumValue;
  }): Enum<TEnumValue> {
    // @ts-expect-error
    return { enumInstance };
  },
};
