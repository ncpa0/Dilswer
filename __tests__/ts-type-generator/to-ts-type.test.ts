import { ExternalTypeImport } from "@TsTypeGenerator/parser-options";
import { getMetadata, OptionalField, toTsType, Type } from "../../src/index";

enum Enum {
  A = "A",
  B = "B",
  C = "C",
}

const testDt = Type.Record({
  uuid: Type.String.setDescription("User unique identifier"),
  number: Type.Number,
  bool: Type.Boolean,
  stringSet: Type.Set(Type.String),
  stringArray: Type.Array(Type.String),
  unionArray: Type.Array(
    Type.Record({ foo: Type.String }),
    Type.Record({ bar: Type.String }),
  ).setDescription("Array of two possible record types"),
  tuple: Type.Tuple(Type.String, Type.Number),
  namedTuple: Type.Tuple(
    Type.Array(Type.String),
    Type.Enum(Enum).setEnumName("Enum"),
  )
    .setTitle("A Tuple")
    .setDescription("A tuple with a named type"),
  literalString: Type.Literal("literal"),
  literalNumber: Type.Literal(1),
  literalBoolean: Type.Literal(true),
  enum: Type.Enum(Enum).setEnumName("Enum"),
  bMember: Type.EnumMember(Enum.B).setEnumName("Enum").setMemberName("B"),
  recordIntersection: Type.AllOf(
    Type.Record({
      a: Type.String,
    }),
    Type.Record({
      b: Type.Number,
    }),
  ),
  dict: Type.Dict(Type.String),
  oneof: Type.OneOf(
    Type.Record({ foo: Type.String }).setTitle("Foo Container"),
    Type.Record({ bar: Type.String }).setTitle("Bar Container"),
  ),
  symbol: Type.Symbol,
  undef: Type.Undefined,
  customValidator: Type.Custom((value: any): value is any => true),
  stringMatching: Type.StringMatching(/^foo$/),
  namedStringMatching: Type.StringMatching(/^'foo'\..+$/)
    .setTitle("Foo Matcher")
    .setTsPattern("'foo'.${string}"),
  optionalSelfCopy: {
    required: false,
    type: Type.Record({
      id: { required: false, type: Type.String },
      number: { required: false, type: Type.Number },
      bool: { required: false, type: Type.Boolean },
      stringSet: { required: false, type: Type.Set(Type.String) },
      stringArray: {
        required: false,
        type: Type.Array(Type.String),
      },
      unionArray: {
        required: false,
        type: Type.Array(
          Type.Record({ foo: Type.String }).setTitle("Foo Container"),
          Type.Record({ bar: Type.String }),
        ),
      },
      tuple: {
        required: false,
        type: Type.Tuple(Type.String, Type.Number),
      },
      literalString: { required: false, type: Type.Literal("literal") },
      literalNumber: { required: false, type: Type.Literal(1) },
      literalBoolean: { required: false, type: Type.Literal(true) },
      enum: {
        required: false,
        type: Type.Enum(Enum).setEnumName("Enum"),
      },
      bMember: {
        required: false,
        type: Type.EnumMember(Enum.B)
          .setEnumName("Enum")
          .setMemberName("B"),
      },
      recordIntersection: {
        required: false,
        type: Type.AllOf(
          Type.Record({
            a: Type.String,
          }),
          Type.Record({
            b: Type.Number,
          }),
        ).setTitle("SomeIntersection"),
      },
      dict: { required: false, type: Type.Dict(Type.String) },
      oneof: {
        required: false,
        type: Type.OneOf(
          Type.Record({ foo: Type.String }),
          Type.Record({ bar: Type.String }),
        ),
      },
    })
      .setTitle("Optional Test Record")
      .setDescription(
        "A record identical to the parent record, but with all fields optional",
      ),
  },
}).setTitle("Test Record");

describe("toTsType", () => {
  it("should correctly generate a complex type with 'compact' mode", () => {
    const tsType = toTsType(testDt);

    expect(tsType).toMatchSnapshot();
  });

  it("should correctly generate a complex type with 'named-expanded' mode", () => {
    const tsType = toTsType(testDt, {
      mode: "named-expanded",
      onDuplicateName: "rename",
    });

    expect(tsType).toMatchSnapshot();
  });

  it("should correctly generate a complex type with 'fully-expanded' mode", () => {
    const tsType = toTsType(testDt, {
      mode: "fully-expanded",
      onDuplicateName: "rename",
    });

    expect(tsType).toMatchSnapshot();
  });

  describe("should correctly add export statements", () => {
    describe("when exporting only main", () => {
      it("should not add `declare`", () => {
        const dt = Type.Record({
          foo: Type.Array(Type.String).setTitle("FooList"),
        }).setTitle("Main");

        const tsType = toTsType(dt, {
          declaration: false,
          exports: "main",
          mode: "named-expanded",
          onDuplicateName: "rename",
        });

        expect(tsType).toMatchSnapshot();
      });

      it("should add `declare`", () => {
        const dt = Type.Record({
          foo: Type.Array(Type.String).setTitle("FooList"),
        }).setTitle("Main");

        const tsType = toTsType(dt, {
          declaration: true,
          exports: "main",
          mode: "named-expanded",
          onDuplicateName: "rename",
        });

        expect(tsType).toMatchSnapshot();
      });
    });

    describe("when exporting only named", () => {
      it("should not add `declare`", () => {
        const dt = Type.Record({
          unnamed: Type.Array(Type.Number),
          foo: Type.Array(Type.String).setTitle("FooList"),
        }).setTitle("Main");

        const tsType = toTsType(dt, {
          declaration: false,
          exports: "named",
          mode: "fully-expanded",
          onDuplicateName: "rename",
        });

        expect(tsType).toMatchSnapshot();
      });

      it("should add `declare`", () => {
        const dt = Type.Record({
          unnamed: Type.Array(Type.Number),
          foo: Type.Array(Type.String).setTitle("FooList"),
        }).setTitle("Main");

        const tsType = toTsType(dt, {
          declaration: true,
          exports: "named",
          mode: "fully-expanded",
          onDuplicateName: "rename",
        });

        expect(tsType).toMatchSnapshot();
      });
    });

    describe("when exporting all", () => {
      it("should not add `declare`", () => {
        const dt = Type.Record({
          unnamed: Type.Array(Type.Number),
          foo: Type.Array(Type.String).setTitle("FooList"),
        }).setTitle("Main");

        const tsType = toTsType(dt, {
          declaration: false,
          exports: "all",
          mode: "fully-expanded",
          onDuplicateName: "rename",
        });

        expect(tsType).toMatchSnapshot();
      });

      it("should add `declare`", () => {
        const dt = Type.Record({
          unnamed: Type.Array(Type.Number),
          foo: Type.Array(Type.String).setTitle("FooList"),
        }).setTitle("Main");

        const tsType = toTsType(dt, {
          declaration: true,
          exports: "all",
          mode: "fully-expanded",
          onDuplicateName: "rename",
        });

        expect(tsType).toMatchSnapshot();
      });
    });
  });

  describe("should correctly generate a type referencing a class", () => {
    it("in compact mode with global FooBar", () => {
      class FooBar {}

      const dt = Type.Record({
        foobar: Type.InstanceOf(FooBar),
      });

      const tsType = toTsType(dt);

      expect(tsType).toMatchSnapshot();
    });

    it("in fully-expanded mode with global FooBar", () => {
      class FooBar {}

      const dt = Type.Record({
        foobar: Type.InstanceOf(FooBar),
      });

      const tsType = toTsType(dt, { mode: "fully-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("in fully-expanded mode with global FooBar", () => {
      class FooBar {}

      const dt = Type.Record({
        foobar: Type.InstanceOf(FooBar).setTitle("FooBar"),
      });

      const tsType = toTsType(dt, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("in compact mode with imported FooBar", () => {
      class FooBar {}

      const dt = Type.Record({
        foobar: Type.InstanceOf(FooBar),
      });

      const tsType = toTsType(dt, {
        getExternalTypeImport(type) {
          if (type.kind === "instanceOf" && type.instanceOf === FooBar) {
            return {
              path: "./foo-bar",
              typeName: "FooBar",
            };
          }
        },
      });

      expect(tsType).toMatchSnapshot();
    });

    it("in fully-expanded mode with imported FooBar", () => {
      class FooBar {}

      const dt = Type.Record({
        foobar: Type.InstanceOf(FooBar),
      });

      const tsType = toTsType(dt, {
        mode: "fully-expanded",
        getExternalTypeImport(type) {
          if (type.kind === "instanceOf" && type.instanceOf === FooBar) {
            return {
              path: "./foo-bar",
              typeName: "FooBar",
            };
          }
        },
      });

      expect(tsType).toMatchSnapshot();
    });

    it("in fully-expanded mode with imported FooBar", () => {
      class FooBar {}

      const dt = Type.Record({
        foobar: Type.InstanceOf(FooBar).setTitle("FooBar"),
      });

      const tsType = toTsType(dt, {
        mode: "named-expanded",
        getExternalTypeImport(type) {
          if (type.kind === "instanceOf" && type.instanceOf === FooBar) {
            return {
              path: "./foo-bar",
              typeName: "FooBar",
            };
          }
        },
      });

      expect(tsType).toMatchSnapshot();
    });
  });

  describe("should correctly add external type import statements", () => {
    it("with regular Custom validator", () => {
      const type = Type.Record({
        foo: Type.Custom(
          (v): v is string => typeof v === "string",
        ).setExtra(
          {
            typeName: "CustomFoo",
            path: "./custom-validator.d.ts",
          } satisfies ExternalTypeImport,
        ),
      });

      const tsType = toTsType(type, {
        mode: "named-expanded",
        getExternalTypeImport(type) {
          const meta = getMetadata<ExternalTypeImport>(type);
          if (meta.extra) {
            return meta.extra;
          }
        },
      });

      expect(tsType).toMatchSnapshot();
    });

    it("with regular Enum", () => {
      const type = Type.Record({
        foo: Type.Enum(Enum).setExtra(
          {
            typeName: "MyEnum",
            path: "./enum.d.ts",
          } satisfies ExternalTypeImport,
        ),
      });

      const tsType = toTsType(type, {
        mode: "named-expanded",
        getExternalTypeImport(type) {
          const meta = getMetadata<ExternalTypeImport>(type);
          if (meta.extra) {
            return meta.extra;
          }
        },
      });

      expect(tsType).toMatchSnapshot();
    });

    it("with regular Enum Member", () => {
      const type = Type.Record({
        foo: Type.EnumMember(Enum.C)
          .setMemberName("C")
          .setExtra(
            {
              typeName: "MyEnum",
              path: "./enum.d.ts",
            } satisfies ExternalTypeImport,
          ),
      });

      const tsType = toTsType(type, {
        mode: "named-expanded",
        getExternalTypeImport(type) {
          const meta = getMetadata<ExternalTypeImport>(type);
          if (meta.extra) {
            return meta.extra;
          }
        },
      });

      expect(tsType).toMatchSnapshot();
    });

    it("with regular Function as a type", () => {
      const type = Type.Record({
        foo: Type.Function.setExtra(
          {
            typeName: "MyFunction",
            path: "./my-function.d.ts",
          } satisfies ExternalTypeImport,
        ),
      });

      const tsType = toTsType(type, {
        mode: "named-expanded",
        getExternalTypeImport(type) {
          const meta = getMetadata<ExternalTypeImport>(type);
          if (meta.extra) {
            return meta.extra;
          }
        },
      });

      expect(tsType).toMatchSnapshot();
    });

    it("with regular Function as a value", () => {
      const type = Type.Record({
        foo: Type.Function.setExtra(
          {
            typeName: "myFunction",
            path: "./my-function.d.ts",
            valueImport: true,
          } satisfies ExternalTypeImport,
        ),
      });

      const tsType = toTsType(type, {
        mode: "named-expanded",
        getExternalTypeImport(type) {
          const meta = getMetadata<ExternalTypeImport>(type);
          if (meta.extra) {
            return meta.extra;
          }
        },
      });

      expect(tsType).toMatchSnapshot();
    });

    it("and export it directly if it's a root type", () => {
      const foo = Type.Custom((v): v is string => typeof v === "string")
        .setTitle("IsStringFn")
        .setExtra(
          {
            typeName: "IsStringFn",
            path: "./is-string-fn.d.ts",
          } satisfies ExternalTypeImport,
        );

      const tsType = toTsType(foo, {
        mode: "named-expanded",
        getExternalTypeImport(type) {
          const meta = getMetadata<ExternalTypeImport>(type);
          if (meta.extra) {
            return meta.extra;
          }
        },
      });

      expect(tsType).toMatchSnapshot();
    });
  });

  describe("should work correctly with circular type", () => {
    it("simple circular record", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Array(self),
        })
      );

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("simple circular named record", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Array(self),
        }).setTitle("Node")
      );

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("immediately self-referencing record", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          self: OptionalField(self),
        }).setTitle("SelfReferencingRecord")
      );

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("simple circular array", () => {
      const typeDef = Type.Recursive((self) => Type.Array(Type.String, self));

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("deeply nested structure", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Record({
            type: Type.String,
            container: Type.Record({
              name: Type.String,
              children: Type.Array(self),
            }),
          }),
        }).setTitle("Node")
      );

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("deeply nested structure of named types", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Array(
            Type.Record({
              name: Type.Literal("span"),
              children: Type.Array(self),
            }).setTitle("SpanNode"),
          ).setTitle("SpanNodeList"),
        }).setTitle("Node")
      );

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("deeply nested circular types", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Recursive((self2) =>
            Type.Array(
              self,
              Type.Record({
                nested: Type.Literal(true),
                items: self2,
              }),
            )
          ),
        })
      );

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("works with sets", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Set(self),
        })
      );

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("works with direct sets", () => {
      const typeDef = Type.Recursive((self) => Type.Set(self, Type.String));

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("works with dicts", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Dict(self),
        })
      );

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("works with direct dicts", () => {
      const typeDef = Type.Recursive((self) => Type.Dict(self, Type.String));

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("works with tuples", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: OptionalField(
            Type.OneOf(
              Type.Tuple(self),
              Type.Tuple(self, self),
              Type.Tuple(self, self, self),
            ),
          ),
        })
      );

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("works with direct tuples", () => {
      const typeDef = Type.Recursive((self) => Type.Tuple(self, Type.String));

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("works with unions", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: OptionalField(
            Type.OneOf(
              self,
              Type.Record({
                name: Type.String,
                children: self,
              }),
            ),
          ),
        })
      );

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("works with direct unions", () => {
      const typeDef = Type.Recursive((self) => Type.OneOf(self, Type.String));

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });

    it("works with intersections", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: OptionalField(
            Type.AllOf(
              self,
              Type.Record({
                type: Type.String,
              }),
            ),
          ),
        })
      );

      const tsType = toTsType(typeDef, { mode: "named-expanded" });

      expect(tsType).toMatchSnapshot();
    });
  });
});
