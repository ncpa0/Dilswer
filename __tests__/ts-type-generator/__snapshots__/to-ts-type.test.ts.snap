// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toTsType should correctly add export statements when exporting all should add \`declare\` 1`] = `
"export declare type ArrayType = Array<number>;


export declare type FooList = Array<string>;


export declare type Main = {
  unnamed: ArrayType;
  foo: FooList;
};
"
`;

exports[`toTsType should correctly add export statements when exporting all should not add \`declare\` 1`] = `
"export type ArrayType = Array<number>;


export type FooList = Array<string>;


export type Main = {
  unnamed: ArrayType;
  foo: FooList;
};
"
`;

exports[`toTsType should correctly add export statements when exporting only main should add \`declare\` 1`] = `
"declare type FooList = Array<string>;


declare type Main = {
  foo: FooList;
};

export type { Main };
"
`;

exports[`toTsType should correctly add export statements when exporting only main should not add \`declare\` 1`] = `
"type FooList = Array<string>;


type Main = {
  foo: FooList;
};

export type { Main };
"
`;

exports[`toTsType should correctly add export statements when exporting only named should add \`declare\` 1`] = `
"declare type ArrayType = Array<number>;


export declare type FooList = Array<string>;


export declare type Main = {
  unnamed: ArrayType;
  foo: FooList;
};
"
`;

exports[`toTsType should correctly add export statements when exporting only named should not add \`declare\` 1`] = `
"type ArrayType = Array<number>;


export type FooList = Array<string>;


export type Main = {
  unnamed: ArrayType;
  foo: FooList;
};
"
`;

exports[`toTsType should correctly add external type import statements and export it directly if it's a root type 1`] = `
"import type { IsStringFn } from "./is-string-fn.d.ts";
export type { IsStringFn };
"
`;

exports[`toTsType should correctly add external type import statements with regular Custom validator 1`] = `
"import type { CustomFoo } from "./custom-validator.d.ts";

export type RecordType = {
  foo: CustomFoo;
};
"
`;

exports[`toTsType should correctly add external type import statements with regular Enum 1`] = `
"import type { MyEnum } from "./enum.d.ts";

export type RecordType = {
  foo: MyEnum;
};
"
`;

exports[`toTsType should correctly add external type import statements with regular Enum Member 1`] = `
"import type { MyEnum } from "./enum.d.ts";

export type RecordType = {
  foo: MyEnum.C;
};
"
`;

exports[`toTsType should correctly add external type import statements with regular Function as a type 1`] = `
"import type { MyFunction } from "./my-function.d.ts";

export type RecordType = {
  foo: MyFunction;
};
"
`;

exports[`toTsType should correctly add external type import statements with regular Function as a value 1`] = `
"import type { myFunction } from "./my-function.d.ts";

export type RecordType = {
  foo: typeof myFunction;
};
"
`;

exports[`toTsType should correctly generate a complex type with 'compact' mode 1`] = `
"export type TestRecord = {
  /**
   * User unique identifier
   */
  uuid: string;
  number: number;
  bool: boolean;
  stringSet: Set<string>;
  stringArray: Array<string>;
  /**
   * Array of two possible record types
   */
  unionArray: Array<{
    foo: string;
  } | {
    bar: string;
  }>;
  tuple: [string, number];
  /**
   * A tuple with a named type
   */
  namedTuple: [Array<string>, Enum];
  literalString: "literal";
  literalNumber: 1;
  literalBoolean: true;
  enum: Enum;
  bMember: Enum.B;
  recordIntersection: {
    a: string;
  } & {
    b: number;
  };
  dict: Record<string | number, string>;
  oneof: {
    foo: string;
  } | {
    bar: string;
  };
  symbol: symbol;
  undef: undefined;
  customValidator: unknown;
  stringMatching: string;
  namedStringMatching: '\\'foo\\'.\${string}';
  /**
   * A record identical to the parent record, but with all fields optional
   */
  optionalSelfCopy?: {
    id?: string;
    number?: number;
    bool?: boolean;
    stringSet?: Set<string>;
    stringArray?: Array<string>;
    unionArray?: Array<{
      foo: string;
    } | {
      bar: string;
    }>;
    tuple?: [string, number];
    literalString?: "literal";
    literalNumber?: 1;
    literalBoolean?: true;
    enum?: Enum;
    bMember?: Enum.B;
    recordIntersection?: {
      a: string;
    } & {
      b: number;
    };
    dict?: Record<string | number, string>;
    oneof?: {
      foo: string;
    } | {
      bar: string;
    };
  };
};
"
`;

exports[`toTsType should correctly generate a complex type with 'fully-expanded' mode 1`] = `
"type SetType = Set<string>;


type ArrayType = Array<string>;


type RecordType = {
  foo: string;
};


type RecordType1 = {
  bar: string;
};

/**
 * Array of two possible record types
 */
type ArrayType1 = Array<RecordType | RecordType1>;


type Tuple = [string, number];


type ArrayType2 = Array<string>;

/**
 * A tuple with a named type
 */
type ATuple = [ArrayType2, Enum];


type RecordType2 = {
  a: string;
};


type RecordType3 = {
  b: number;
};


type Intersection = RecordType2 & RecordType3;


type Dict = Record<string | number, string>


type FooContainer = {
  foo: string;
};


type BarContainer = {
  bar: string;
};


type Union = FooContainer | BarContainer;


type StringMatching = string;


type FooMatcher = '\\'foo\\'.\${string}';


type SetType1 = Set<string>;


type ArrayType3 = Array<string>;


type FooContainer = {
  foo: string;
};


type RecordType4 = {
  bar: string;
};


type ArrayType4 = Array<FooContainer | RecordType4>;


type Tuple1 = [string, number];


type RecordType5 = {
  a: string;
};


type RecordType6 = {
  b: number;
};


type SomeIntersection = RecordType5 & RecordType6;


type Dict1 = Record<string | number, string>


type RecordType7 = {
  foo: string;
};


type RecordType8 = {
  bar: string;
};


type Union1 = RecordType7 | RecordType8;

/**
 * A record identical to the parent record, but with all fields optional
 */
type OptionalTestRecord = {
  id?: string;
  number?: number;
  bool?: boolean;
  stringSet?: SetType1;
  stringArray?: ArrayType3;
  unionArray?: ArrayType4;
  tuple?: Tuple1;
  literalString?: "literal";
  literalNumber?: 1;
  literalBoolean?: true;
  enum?: Enum;
  bMember?: Enum.B;
  recordIntersection?: SomeIntersection;
  dict?: Dict1;
  oneof?: Union1;
};


type TestRecord = {
  /**
   * User unique identifier
   */
  uuid: string;
  number: number;
  bool: boolean;
  stringSet: SetType;
  stringArray: ArrayType;
  /**
   * Array of two possible record types
   */
  unionArray: ArrayType1;
  tuple: Tuple;
  /**
   * A tuple with a named type
   */
  namedTuple: ATuple;
  literalString: "literal";
  literalNumber: 1;
  literalBoolean: true;
  enum: Enum;
  bMember: Enum.B;
  recordIntersection: Intersection;
  dict: Dict;
  oneof: Union;
  symbol: symbol;
  undef: undefined;
  customValidator: unknown;
  stringMatching: StringMatching;
  namedStringMatching: FooMatcher;
  /**
   * A record identical to the parent record, but with all fields optional
   */
  optionalSelfCopy?: OptionalTestRecord;
};

export type { TestRecord };
"
`;

exports[`toTsType should correctly generate a complex type with 'named-expanded' mode 1`] = `
"/**
 * A tuple with a named type
 */
type ATuple = [Array<string>, Enum];


type FooContainer = {
  foo: string;
};


type BarContainer = {
  bar: string;
};


type FooMatcher = '\\'foo\\'.\${string}';


type FooContainer = {
  foo: string;
};


type SomeIntersection = {
  a: string;
} & {
  b: number;
};

/**
 * A record identical to the parent record, but with all fields optional
 */
type OptionalTestRecord = {
  id?: string;
  number?: number;
  bool?: boolean;
  stringSet?: Set<string>;
  stringArray?: Array<string>;
  unionArray?: Array<FooContainer | {
    bar: string;
  }>;
  tuple?: [string, number];
  literalString?: "literal";
  literalNumber?: 1;
  literalBoolean?: true;
  enum?: Enum;
  bMember?: Enum.B;
  recordIntersection?: SomeIntersection;
  dict?: Record<string | number, string>;
  oneof?: {
    foo: string;
  } | {
    bar: string;
  };
};


type TestRecord = {
  /**
   * User unique identifier
   */
  uuid: string;
  number: number;
  bool: boolean;
  stringSet: Set<string>;
  stringArray: Array<string>;
  /**
   * Array of two possible record types
   */
  unionArray: Array<{
    foo: string;
  } | {
    bar: string;
  }>;
  tuple: [string, number];
  /**
   * A tuple with a named type
   */
  namedTuple: ATuple;
  literalString: "literal";
  literalNumber: 1;
  literalBoolean: true;
  enum: Enum;
  bMember: Enum.B;
  recordIntersection: {
    a: string;
  } & {
    b: number;
  };
  dict: Record<string | number, string>;
  oneof: FooContainer | BarContainer;
  symbol: symbol;
  undef: undefined;
  customValidator: unknown;
  stringMatching: string;
  namedStringMatching: FooMatcher;
  /**
   * A record identical to the parent record, but with all fields optional
   */
  optionalSelfCopy?: OptionalTestRecord;
};

export type { TestRecord };
"
`;

exports[`toTsType should correctly generate a type referencing a class in compact mode with global FooBar 1`] = `
"export type RecordType = {
  foobar: InstanceType<typeof FooBar>;
};
"
`;

exports[`toTsType should correctly generate a type referencing a class in compact mode with imported FooBar 1`] = `
"import type { FooBar } from "./foo-bar";

export type RecordType = {
  foobar: InstanceType<typeof FooBar>;
};
"
`;

exports[`toTsType should correctly generate a type referencing a class in fully-expanded mode with global FooBar 1`] = `
"type ClassInstance = InstanceType<typeof FooBar>


type RecordType = {
  foobar: ClassInstance;
};

export type { RecordType };
"
`;

exports[`toTsType should correctly generate a type referencing a class in fully-expanded mode with global FooBar 2`] = `
"type AliasFooBar = InstanceType<typeof FooBar>


export type RecordType = {
  foobar: AliasFooBar;
};
"
`;

exports[`toTsType should correctly generate a type referencing a class in fully-expanded mode with imported FooBar 1`] = `
"import type { FooBar } from "./foo-bar";

type ClassInstance = InstanceType<typeof FooBar>


type RecordType = {
  foobar: ClassInstance;
};

export type { RecordType };
"
`;

exports[`toTsType should correctly generate a type referencing a class in fully-expanded mode with imported FooBar 2`] = `
"import type { FooBar } from "./foo-bar";

type AliasFooBar = InstanceType<typeof FooBar>


export type RecordType = {
  foobar: AliasFooBar;
};
"
`;

exports[`toTsType should work correctly with circular type deeply nested circular types 1`] = `
"type ArrayType = Array<RecordType | {
  nested: true;
  items: ArrayType;
}>;


type RecordType = {
  name: string;
  children: ArrayType;
};

export type { RecordType };
"
`;

exports[`toTsType should work correctly with circular type deeply nested structure 1`] = `
"type Node = {
  name: string;
  children: {
    type: string;
    container: {
      name: string;
      children: Array<Node>;
    };
  };
};

export type { Node };
"
`;

exports[`toTsType should work correctly with circular type deeply nested structure of named types 1`] = `
"type SpanNode = {
  name: "span";
  children: Array<Node>;
};


type SpanNodeList = Array<SpanNode>;


type Node = {
  name: string;
  children: SpanNodeList;
};

export type { Node };
"
`;

exports[`toTsType should work correctly with circular type immediately self-referencing record 1`] = `
"type SelfReferencingRecord = {
  name: string;
  self?: SelfReferencingRecord;
};

export type { SelfReferencingRecord };
"
`;

exports[`toTsType should work correctly with circular type simple circular array 1`] = `
"type ArrayType = Array<string | ArrayType>;

export type { ArrayType };
"
`;

exports[`toTsType should work correctly with circular type simple circular named record 1`] = `
"type Node = {
  name: string;
  children: Array<Node>;
};

export type { Node };
"
`;

exports[`toTsType should work correctly with circular type simple circular record 1`] = `
"type RecordType = {
  name: string;
  children: Array<RecordType>;
};

export type { RecordType };
"
`;

exports[`toTsType should work correctly with circular type works with dicts 1`] = `
"type RecordType = {
  name: string;
  children: Record<string | number, RecordType>;
};

export type { RecordType };
"
`;

exports[`toTsType should work correctly with circular type works with direct dicts 1`] = `
"type Dictionary = Record<string | number, Dictionary | string>

export type { Dictionary };
"
`;

exports[`toTsType should work correctly with circular type works with direct sets 1`] = `
"type SetType = Set<SetType | string>;

export type { SetType };
"
`;

exports[`toTsType should work correctly with circular type works with direct tuples 1`] = `
"type Tuple = [Tuple, string];

export type { Tuple };
"
`;

exports[`toTsType should work correctly with circular type works with direct unions 1`] = `
"type OneOf = OneOf | string;

export type { OneOf };
"
`;

exports[`toTsType should work correctly with circular type works with intersections 1`] = `
"type RecordType = {
  name: string;
  children?: RecordType & {
    type: string;
  };
};

export type { RecordType };
"
`;

exports[`toTsType should work correctly with circular type works with sets 1`] = `
"type RecordType = {
  name: string;
  children: Set<RecordType>;
};

export type { RecordType };
"
`;

exports[`toTsType should work correctly with circular type works with tuples 1`] = `
"type RecordType = {
  name: string;
  children?: [RecordType] | [RecordType, RecordType] | [RecordType, RecordType, RecordType];
};

export type { RecordType };
"
`;

exports[`toTsType should work correctly with circular type works with unions 1`] = `
"type RecordType = {
  name: string;
  children?: RecordType | {
    name: string;
    children: RecordType;
  };
};

export type { RecordType };
"
`;
